group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -10225,53 +10251,61 @@ public void testConsumerGroupHeartbeatToClassicGroupFromExistingStaticMember() {
+
+        // The static member rejoins the new consumer group.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupMemberSubscriptionRecord(groupId, expectedFinalConsumerMember));
+
+        // The subscription metadata hasn't been updated during the conversion, so a new one is computed.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupSubscriptionMetadataRecord(groupId,
+            Map.of(fooTopicName, new TopicMetadata(fooTopicId, fooTopicName, 1))));
+
+        // Newly joining static member bumps the group epoch. A new target assignment is computed.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupEpochRecord(groupId, 1));
+
+        // If the memberId is generated by the consumer itself, the consumer should retain this memberId.
+        // As a result, the record won't contain a new target assignment record.
+        // If the memberId is not consumer-generated, add a new target assignment record to the expected records,
+        // since a different memberId will be considered as a new member.
+        if (!isConsumerGeneratedMemberId) {
+            expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentRecord(groupId, newMemberId,
+                mkAssignment(mkTopicAssignment(fooTopicId, 0))));
+        }
 
-            // Newly joining static member bumps the group epoch. A new target assignment is computed.
-            GroupCoordinatorRecordHelpers.newConsumerGroupEpochRecord(groupId, 1),
-            GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentRecord(groupId, newMemberId, mkAssignment(mkTopicAssignment(fooTopicId, 0))),
-            GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentEpochRecord(groupId, 1),
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentEpochRecord(groupId, 1));
 
-            // The newly created static member takes the assignment from the existing member.
-            // Bump its member epoch and transition to STABLE.
-            GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, expectedFinalConsumerMember)
-        );
+        // The newly created static member takes the assignment from the existing member.
+        // Bump its member epoch and transition to STABLE.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, expectedFinalConsumerMember));
 
         assertRecordsEquals(expectedRecords, result.records());
         context.assertSessionTimeout(groupId, newMemberId, 45000);
     }
 
-    @Test
-    public void testConsumerGroupHeartbeatFromExistingClassicStaticMember() {
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testConsumerGroupHeartbeatFromExistingClassicStaticMember(boolean isConsumerGeneratedMemberId) {
```

Comment:
We should perhaps use the last version in this test and generate the member id on the client side.